### DISCUSSION TITLE
Two button both pressed (sample LongPress)

### DISCUSSION THREAD

**Discussion by ysoyipek** [bxparks/AceButton#96] [General]
Hi, congratulations on a very nice project. 

What I want to ask is, is it possible to have a feature that I think would be very nice if added?

I need to use the long-press method of two keys to turn a device off and on safely.

I am able to check and catch the isPressedRaw() property on both keys in the kEventLongPressed event. But the event throwing occurs on the second depressed key.

I wonder if it's possible to do two-key presses with the existing library?

code;
```
#include <AceButton.h>
using namespace ace_button;

#define PIN_BUTTON1 6
#define PIN_BUTTON2 7

AceButton button1(PIN_BUTTON1);
AceButton button2(PIN_BUTTON2);

void handleEvent(AceButton *, uint8_t, uint8_t);

void setup()
{
  pinMode(PIN_BUTTON1, INPUT_PULLUP);
  pinMode(PIN_BUTTON2, INPUT_PULLUP);

  ButtonConfig *buttonConfig = ButtonConfig::getSystemButtonConfig();
  buttonConfig->setEventHandler(handleEvent);
  buttonConfig->setFeature(ButtonConfig::kFeatureLongPress);
  buttonConfig->setLongPressDelay(3000);
}

void handleEvent(AceButton *button, uint8_t eventType, uint8_t buttonState)
{
  switch (eventType)
  {
  case AceButton::kEventLongPressed:
    if (button1.isPressedRaw() && button2.isPressedRaw())
    {
      Serial.println("LongPressed: Both");
    }
    break;
  }
}
```

terminal;
```
LongPressed: Both
LongPressed: Both
```

**Comment by bxparks** [comment]
There have been a handful of previous questions about supporting simultaneous buttons, see for example:

* https://github.com/bxparks/AceButton/discussions/94
* https://github.com/bxparks/AceButton/discussions/83

The library does not support multiple buttons directly. My inclination so far has been to push the responsibility for handling multiple buttons into the EventHandler or IEventHandler provided by the developer of the application code. My rationale is that:

1) I don't personally use multiple buttons, so I have not spent a lot of time and effort thinking about the problem space.
2) My intuition tells me that the state machine for handling all the various interactions of events from 2 buttons is *far far* more complex than the state machine for a single button, and I won't have the time and energy to handle the general case.
3) Therefore, I delegate the responsibility for handling multiple buttons to the application developer who is expected to write the custom EventHandler or IEventHandler that handles the multiple buttons.

In your particular example, you will need to think about and design the specification of exactly how you want your button combinations to behave:

1) Do the buttons need to be pressed at the same time, or can they be slightly off by a few milliseconds. How many milliseconds?
2) If one button is pressed down, then a LongPressed is triggered, then the other button is pressed down and a LongPressed triggered, does that count as a DoubleButtonLongPressed?
3) If both buttons are DoubleButtonLongPressed, and you release one of them, then press it again, does it trigger another DoubleButtonLongPressed? It depends on what you want.
4) If both buttons are DoubleButtonLongPressed for than some amount of time, do you then want to trigger DoubleButtonRepeatPressed? I don't know, it will depend.
5) If both buttons are DoubleButtonLongPressed, and you lift one of the buttons, do you want the remaining pressed button to continue to trigger the LongPressed event? I don't know.

As you can see, everything depends on what you want your app and your buttons to do. I don't think I have the capacity to design a configuration mechanism that can support everything that anyone would ever want to happen with the combination of 2 buttons.

So take a look in https://github.com/bxparks/AceButton/discussions/83 for the  look at my [examples/SimultaneousButtons](https://github.com/bxparks/AceButton/tree/develop/examples/SimultaneousButtons). You will want to create your own subclass of `IEventHandler` that intercepts the events generated by your 2 buttons, then generate the synthetic DoubleButtonLongPressed events that you want to handle.

I have limited time to spend on this library, so I hope you can do most of the heavy lifting yourself with the hints that I have given above. Good luck!
